Add LTC3220 driver

edelweiseescala opened the pull request:
## PR Description

- Add LTC3220 driver and documentation.
- Datasheet: [LTC3220](https://www.analog.com/media/en/technical-documentation/data-sheets/32201fd.pdf)
- Tested on RPI4 using the following eval boards: [DC1265A-A](https://www.analog.com/en/resources/evaluation-hardware-and-software/evaluation-boards-kits/dc1265a-a.html), [DC1265A-B](https://www.analog.com/en/resources/evaluation-hardware-and-software/evaluation-boards-kits/dc1265a-b.html)

## PR Type
- [ ] Bug fix (a change that fixes an issue)
- [x] New feature (a change that adds new functionality)
- [ ] Breaking change (a change that affects other repos or cause CIs to fail)

## PR Checklist
- [x] I have conducted a self-review of my own code changes
- [x] I have compiled my changes, including the documentation
- [x] I have tested the changes on the relevant hardware
- [ ] I have updated the documentation outside this repo accordingly
- [ ] I have provided links for the relevant upstream lore

artursartamonovsadi commented on drivers/leds/leds-ltc3220.c:
```diff
@@ -0,0 +1,650 @@
+}
+
+static ssize_t ltc3220_force_cpo_2x_store(struct device *dev,
+						struct device_attribute *devAttr,
```
same as above

artursartamonovsadi commented on drivers/leds/leds-ltc3220.c:
```diff
@@ -0,0 +1,650 @@
+}
+
+static ssize_t ltc3220_quick_write_store(struct device *dev,
+					struct device_attribute *devAttr,
```
same above

vasbimpikasadi commented on drivers/leds/leds-ltc3220.c:
```diff
@@ -0,0 +1,650 @@
+{
+	int ret;
+	struct i2c_client *client = uled->client;
+	u8 reg_val = (mode << 6) | current_level;
```
This could overflow. Masking can prevent that. e.g. `u8 reg_val = ((mode & 0x03) << 6) | (current_level & 0x3F);`

vasbimpikasadi commented on drivers/leds/leds-ltc3220.c:
```diff
@@ -0,0 +1,650 @@
+	struct ltc3220_uled_cfg *uled_cfg;
+
+	current_level = (int)brightness;
+	if (current_level > MAX_CURRENT)
```
You know the code way better than I do, but shouldn't that be an error? If it's fine to continue with current capped at max, at least a dev_info should be printed?

vasbimpikasadi reviewed (changes requested):
That shows great effort! It's a bit sparse on comments and I've noted some changes which I think have to be fixed before this gets approved.

CalebEthridgeADI commented on include/linux/platform_data/leds-ltc3220.h:
```diff
@@ -0,0 +1,64 @@
+#ifndef __LINUX_I2C_LTC3220_H
+#define __LINUX_I2C_LTC3220_H
+
+#define NUM_LEDS 18
```
Can you add a LTC3220 prefix before these defines like you do in leds-ltc3220.c i.e `LTC3220_NUM_LEDS` ?

nunojsa commented on Documentation/devicetree/bindings/leds/leds-ltc3220.yaml:
```diff
@@ -0,0 +1,168 @@
+  reset-gpios:
+    maxItems: 1
+    description: |
+      GPIO attached to the chip's reset pin
```
Ditto about the description

nunojsa commented on Documentation/devicetree/bindings/leds/leds-ltc3220.yaml:
```diff
@@ -0,0 +1,168 @@
+          - 15 # LED output D15
+          - 16 # LED output D16
+          - 17 # LED output D17
+          - 18 # LED output D18
```
I guess the reg property is `required` in the child node.

nunojsa replied:
Why not minimum = 1 and maximum = 18?

nunojsa commented on Documentation/devicetree/bindings/leds/leds-ltc3220.yaml:
```diff
@@ -0,0 +1,168 @@
+
+             led@18 {
+                 reg = <18>;
+             };
```
Don't put all the child nodes in the example. I would also add a bit more in the node since you're getting properties from common.yaml.

nunojsa commented on drivers/leds/leds-ltc3220.c:
```diff
@@ -0,0 +1,650 @@
+ */
+
+#include <linux/module.h>
+#include <linux/kernel.h>
```
do you need kernel.h?

nunojsa commented on drivers/leds/leds-ltc3220.c:
```diff
@@ -0,0 +1,650 @@
+#include <linux/kernel.h>
+#include <linux/i2c.h>
+#include <linux/leds.h>
+#include <linux/platform_data/leds-ltc3220.h>
```
Fairly sure the above header is to be dropped

nunojsa commented on drivers/leds/leds-ltc3220.c:
```diff
@@ -0,0 +1,650 @@
+#define LTC3220_ULED18      0x12
+#define LTC3220_GRAD_BLINK  0x13
+
+static int ltc3220_write(struct i2c_client *client, u8 reg, u8 val)
```
Agreed. I don't see any real benefit in these wrappers. I would also see if regmap is a possibility

nunojsa commented on drivers/leds/leds-ltc3220.c:
```diff
@@ -0,0 +1,650 @@
+	platform_data->reset_gpio = devm_gpiod_get_optional(&client->dev, "reset", GPIOD_OUT_HIGH);
+	if (IS_ERR(platform_data->reset_gpio)) {
+		dev_err(&client->dev, "Fail to set reset GPIO\n");
+		return -ENOMEM;
```
In fact, `return dev_err_probe(dev, PTR_ERR(platform_data->reset_gpio), ...);`

nunojsa commented on drivers/leds/leds-ltc3220.c:
```diff
@@ -0,0 +1,650 @@
+MODULE_AUTHOR("Edelweise Escala <edelweise.escala@analog.com>");
+MODULE_DESCRIPTION("LED driver for LTC3220 controllers");
+MODULE_LICENSE("GPL");
+MODULE_ALIAS("platform:leds-ltc3220");
```
Typically not used anymore

nunojsa reviewed (commented):
First round from me. This one will need some iterations before going upstream. I'm also not that familiar with the LED subsystem so I might give it a look at some point to be more helpful.

edelweiseescala commented on drivers/leds/leds-ltc3220.c:
```diff
@@ -0,0 +1,650 @@
+	platform_data->led_cdev.groups = ltc3220_config_groups;
+	ret = devm_led_classdev_register_ext(&client->dev,
+							&platform_data->led_cdev,
+							&init_data_config);
```
It is like device level settings

edelweiseescala commented:
V2 Changelog:

dt-bindings: leds: ltc3220: Document LTC3220 18 channel LED Driver
-used adi
-dropped unnecessary descriptions
-dropped | formatting
-fixed led pattern

leds: ltc3220: add driver
-Fix naming which does not follow Linux convention, also fixed naming for defines with onw LTC3220 prefix. Follow kernel style
-Remove unnecessary headers, and fixed header arrangement
-Remove unnecessary functions (ltc3220_write)
-Used FIELD_PREP()
-Use firmware agnostic properties
-Use only GPL2
-Led brightness now uses ltc3220_set_led_data
  Instead of 0-63 current level it is now 0-255 where in
  0-63 - normal LED mode
  64-127 - blink mode
  127-191 - gradation mode
  191-255 - GPO
  this removes the need for another custom ABI for individual led mode.

Added leds: ltc3220: Documentation for Custom ABI
-documentation for the added custom ABI

jemfgeronimo commented on Documentation/devicetree/bindings/leds/leds-ltc3220.yaml:
```diff
@@ -0,0 +1,168 @@
+
+examples:
+  - |
+   #include <dt-bindings/leds/common.h>
```
Example indentation should align with `|`
```suggestion
  - |
    #include <dt-bindings/leds/common.h>

jemfgeronimo commented on Documentation/devicetree/bindings/leds/leds-ltc3220.yaml:
```diff
@@ -9,25 +9,19 @@ title: Analog Devices LTC3220 LED Drivers
-description: |
-  Bindings for the Analog Devices LTC3220 18 channel
-  LED Drivers.
+description: Bindings for the Analog Devices LTC3220 18 channel LED Drivers.
```
The changes on this file should be merged with the previous commit.

machschmitt commented on Documentation/devicetree/bindings/leds/leds-ltc3220.yaml:
```diff
@@ -0,0 +1,114 @@
+             reg = <0x1c>;
+             #address-cells = <1>;
+             #size-cells = <0>;
+             reset-gpios = <&gpio 17 0>;
```
the reset-gpio property can become clearer if the example makes use gpio dt-bindings

```yaml
    #include <dt-bindings/gpio/gpio.h>
...

             reset-gpios = <&gpio 17 GPIO_ACTIVE_LOW>;
```

machschmitt reviewed (commented):
Hi @edelweiseescala,

A few comments from my side.
Besides the code changes, I'd suggest to also update your SoB tag with better wording of your name.

nunojsa commented on Documentation/ABI/testing/sysfs-class-led-driver-ltc3220:
```diff
@@ -0,0 +1,54 @@
+		0 - Enables mode logic to control mode changes based on dropout signal
+		1 - Forces charge pump into 1.5x mode
+		2 - Forces charge pump into 2x mode
+		3 - Forces charge pump into 1x mode
```
Hmm this one seems something HW specific? Is this something that should change at runtime? If I'm right, consider a devicetree attr

nunojsa commented on Documentation/ABI/testing/sysfs-class-led-driver-ltc3220:
```diff
@@ -0,0 +1,54 @@
+		which allows parallel write on all 18 LED registers, when data is being written on
+		REG1. If you write 1 quick_write is enabled, 0 to disable.
+
+What:		/sys/class/leds/<led>/gradation_speed
```
I guess this one could be better named gradation_period_ms. Then you would actually pass the period in milliseconds which is a better API for the user. This is possible because we have three distinct periods which match 3 distinct ramp times.

nunojsa commented on drivers/leds/leds-ltc3220.c:
```diff
@@ -0,0 +1,534 @@
+	struct ltc3220_command_cfg command_cfg;
+	struct ltc3220_uled_cfg uled_cfg[LTC3220_NUM_LEDS];
+	struct ltc3220_grad_cfg grad_cfg;
+	struct gpio_desc *reset_gpio;
```
Typically there no need to save this pointer. Any reason for it?

nunojsa commented on drivers/leds/leds-ltc3220.c:
```diff
@@ -0,0 +1,534 @@
+				ret);
+		}
+
+		i = source - 1; //led node reg/index/address goes from 1 to 18
```
not kernel style comment

nunojsa commented on drivers/leds/leds-ltc3220.c:
```diff
@@ -0,0 +1,534 @@
+		ret = fwnode_property_read_u32(child, "reg", &source);
+		if (ret != 0 || !source || source > LTC3220_NUM_LEDS) {
+			dev_warn(&client->dev, "Couldn't read LED address: %d\n",
+				ret);
```
The above is an error. Fail probe in case reg is wrong or not given

nunojsa commented on drivers/leds/leds-ltc3220.c:
```diff
@@ -0,0 +1,534 @@
+		ltc3220_set_command(ltc3220_state, 0, 0, 1);
+		i2c_smbus_write_byte_data(ltc3220_state->client, LTC3220_ULED01, 0);
+		ltc3220_set_command(ltc3220_state, 0, 0, 0);
+		ltc3220_set_blink_and_gradation(ltc3220_state, 0, 0, 0);
```
no error handling in any of the above calls.

nunojsa commented on drivers/leds/leds-ltc3220.c:
```diff
@@ -0,0 +1,534 @@
+	&dev_attr_gradation_is_increasing.attr,
+	&dev_attr_blink_cfg.attr,
+	NULL
+};
```
Be prepared to justify these attributes... I'm still not sure about it

nunojsa commented on drivers/leds/leds-ltc3220.c:
```diff
@@ -0,0 +1,534 @@
+		ltc3220_gradation_is_increasing_store);
+
+static ssize_t ltc3220_blink_cfg_show(struct device *dev,
+								 struct device_attribute *attr, char *buf)
```
Take care about the above

nunojsa commented on drivers/leds/leds-ltc3220.c:
```diff
@@ -0,0 +1,534 @@
+
+	ret = ltc3220_set_blink_and_gradation(ltc3220_state,
+										ltc3220_state->blink_cfg,
+										ltc3220_state->grad_cfg.gradation_speed,
```
same

vasbimpikasadi reviewed (changes requested):
Getting really better, loads of great feedback from everyone. This is shaping up quite nicely.

nunojsa commented on Documentation/devicetree/bindings/leds/leds-ltc3220.yaml:
```diff
@@ -0,0 +1,134 @@
+
+             led@6 {
+                 reg = <6>;
+             };
```
It would be nice to have some of the properties coming from common.yaml that the device supports. Otherwise I wonder why we $ref it.

nunojsa commented on Documentation/ABI/testing/sysfs-class-led-driver-ltc3220:
```diff
@@ -0,0 +1,26 @@
+Contact:	Edelweise Escala <edelweise.escala@analog.com>
+Description:	LTC3220 read only command to show possible gradiation mode such as:
+		"ramp_up"
+		"ramp_down"
```
I'm still not sure about the above ABI. Does it make sense to have these as runtime? I'm tempted to suggest DT but you might see what the maintainer has to say about it. Maybe there's already a way to handle it through the subsystem.

nunojsa commented on drivers/leds/leds-ltc3220.c:
```diff
@@ -0,0 +1,536 @@
+							   &init_data);
+		if (ret) {
+			dev_err(&client->dev, "Fail LED class register\n");
+			return ret;
```
`dev_err_probe()`

nunojsa commented on drivers/leds/leds-ltc3220.c:
```diff
@@ -0,0 +1,536 @@
+	}
+
+	ltc3220_state->command_cfg.is_quick_write =
+		device_property_read_bool(&client->dev, "adi,quick-write-enable");
```
If you have to line break, I would prefer:

```
if (device_property_read_bool(&client->dev, "adi,quick-write-enable"))
    ltc3220_state->command_cfg.is_quick_write = true;
```

edelweiseescala replied:
fixed line break

nunojsa commented on drivers/leds/leds-ltc3220.c:
```diff
@@ -0,0 +1,536 @@
+	if (reset_gpio) {
+		gpiod_set_value_cansleep(reset_gpio, 0);
+		fsleep(20);
+		gpiod_set_value_cansleep(reset_gpio, 1);
```
The above is wrong. You already bring your device in reset (GPIOD_OUT_HIGH). So you just want to bring it out of reset which means

gpiod_set_value_cansleep(reset_gpio, 0);

0 means deassert.  The gpiolib will already properl handle active low pins if you properly set them in DT.

edelweiseescala replied:
Fixed the reset thank you

nunojsa commented on drivers/leds/leds-ltc3220.c:
```diff
@@ -0,0 +1,536 @@
+}
+
+static ssize_t ltc3220_gradation_period_ms_show(struct device *dev,
+								 struct device_attribute *attr, char *buf)
```
Odd indentation

machschmitt reviewed (commented):
Hello, this is my second round of review.

Besides the comments to the code, I'd also suggest to drop the "ltc3220:" part of
"dt-bindings: leds: ltc3220: Document LTC3220 18 channel LED Driver".
ltc3220 doesn't exist in the kernel, yet.

edelweiseescala commented on Documentation/devicetree/bindings/leds/leds-ltc3220.yaml:
```diff
@@ -0,0 +1,129 @@
+    type: boolean
+    description:
+      Enable quick write mode. When enabled, writing to LED D1 (register 0x01)
+      will update all 18 LED outputs with the same brightness value.
```
added a  19th led in device tree with function as quick-write. Thank You!

machschmitt commented on Documentation/devicetree/bindings/leds/leds-ltc3220.yaml:
```diff
@@ -0,0 +1,128 @@
+             led@19 {
+                 reg = <1>;
+                 function = "quick-write";
+             };
```
Makes sense I guess. Having virtual hardware described in dt looks a bit unusual, but I'll leave the final word about it for dt maintainers. Also note address 1 (`reg = <1>`) was already being used by led@1.

machschmitt commented on drivers/leds/leds-ltc3220.c:
```diff
@@ -0,0 +1,502 @@
+		if (is_quick_write_led) {
+			led = &ltc3220_state->quick_write_led;
+			led->led_index = 0;
+			led->led_mode = QUICK_WRITE;
```
I'm a bit concerned about how the "quick_write" led will appear in user space? I mean, is it going to be clear for the user that this led will write to all LTC3220 connected leads? If so, then LGTM.

edelweiseescala replied:
Can I upstream this already?

machschmitt reviewed (commented):
Hi @edelweiseescala,

Driver looks good overall. A few comments left from my side. I'd suggest to wait one week or so (to give time for fellow ADI devs to review this), then send it to the mailing list.

nunojsa commented on drivers/leds/leds-ltc3220.c:
```diff
@@ -0,0 +1,525 @@
+		if (!fwnode_property_read_string(child, "function", &function)) {
+			if (!strcmp(function, "quick-write"))
+				is_quick_write_led = true;
+		}
```
I think the above is a custom property. Isn't quick write affecting all LEDs? If so this should be a device property and it should have a proper vendor prefix. So, a boolean like "adi,quick-write-enable".

Having said the above I'm still not sure where this feature fits the best... Maybe something worth asking upstream when submitting

nunojsa commented on drivers/leds/leds-ltc3220.c:
```diff
@@ -0,0 +1,525 @@
+		ret = fwnode_property_read_u32(child, "reg", &source);
+		if (ret != 0 || !source || source > (LTC3220_NUM_LEDS + 1))
+			return dev_err_probe(&client->dev, ret ? ret : -EINVAL,
+					     "Couldn't read LED address\n");
```
The above is a bit weird and hard to read. Simplify it:

```
ret = fwnode_property_read_u32(...);
if (ret)
    return dev_err_probe(dev, ret, ...);
if (!source || source > ...)
    return dev_err_probe() 
```

nunojsa commented on drivers/leds/leds-ltc3220.c:
```diff
@@ -0,0 +1,525 @@
+	ret = device_property_match_property_string(&client->dev, "adi,force-cpo-level",
+					   ltc3220_cpo_levels, ARRAY_SIZE(ltc3220_cpo_levels));
+	if (ret < 0)
+		ltc3220_state->command_cfg.force_cpo_level = 0;
```
You don't need the above. = is already the default AFAICT. I mean, you're using `devm_kzalloc()`

nunojsa commented on drivers/leds/leds-ltc3220.c:
```diff
@@ -0,0 +1,525 @@
+{
+	struct ltc3220_state *ltc3220_state = i2c_get_clientdata(client);
+
+	ltc3220_shutdown(ltc3220_state);
```
You can use devm_add_action_or_reset() to shutdown the device when unbinding... Then, drop .remove().

nunojsa commented on drivers/leds/leds-ltc3220.c:
```diff
@@ -0,0 +1,525 @@
+	ltc3220_state = uled_cfg->ltc3220_state;
+
+	if (len != 2)
+		return -EINVAL;
```
I would do this before getting `uled_cfg`.

nunojsa commented on drivers/leds/leds-ltc3220.c:
```diff
@@ -0,0 +1,525 @@
+	if (ret < 0) {
+		dev_err(&client->dev, "Failed to set blink/gradation\n");
+		return ret;
+	}
```
Personalty I would simplify and just do `return i2c_smbus_write_byte_data()`

nunojsa reviewed (commented):
Some more comments from me. But we're getting ready for upstream AFAICT

edelweiseescala commented:
V10 Change log

Simplified some if conditions
Remove unnecessary checks
Add missing headers
Remove shutdown remove, reset is enough
fixed spacing issue

Revert to use quick write on device property

nunojsa reviewed (commented):
Just some minor comments. IMO it's ready to be sent upstream

edelweiseescala committed:
leds: Kconfig.adi: imply ltc3220

Add entry for LTC3220 driver

Signed-off-by: Edelweise Escala <edelweise.escala@analog.com>

edelweiseescala committed:
dt-bindings: leds: Add LTC3220 18 channel LED Driver

LTC3220 is a multi-display LED driver with I2C interface.
The LTC3220 provides individual brightness control (64-step),
blinking, and gradation features for up to 18 LED outputs.

Signed-off-by: Edelweise Escala <edelweise.escala@analog.com>
Reviewed-by: Conor Dooley <conor.dooley@microchip.com>
Link: https://patch.msgid.link/20260709-ltc3220-driver-v13-1-9b53237642eb@analog.com
Signed-off-by: Lee Jones <lee@kernel.org>

edelweiseescala committed:
leds: ltc3220: Add Support for LTC3220 18 channel LED Driver

Add driver for the LTC3220 18-channel LED driver
with I2C interface, individual brightness control, and hardware-assisted
blink/gradation features.

Signed-off-by: Edelweise Escala <edelweise.escala@analog.com>
Link: https://patch.msgid.link/20260709-ltc3220-driver-v13-2-9b53237642eb@analog.com
Signed-off-by: Lee Jones <lee@kernel.org>
